VideoPlayer - PVR: fix crash on channel switch for backends not navin…

FernetMenta opened the pull request:
fix trac http://trac.kodi.tv/ticket/16261

FernetMenta committed:
VideoPlayer - PVR: fix crash on channel switch for backends not naving their own demuxer

opdenkamp commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamPVRManager.cpp:
```diff
@@ -383,9 +387,7 @@ bool CDVDInputStreamPVRManager::CloseAndOpen(const char* strFile)
-  PVR_CLIENT client;
-  return g_PVRClients->GetPlayingClient(client) &&
-         client->HandlesInputStream();
+  return m_isOtherStreamHack;
```
just bumped into this PR while checking something else on github ;-) but why don't you just `return m_pOtherStream != NULL;`? looks like you don't need the extra bool for this.

also, this should really be split up into 3 commits (rename method, deleting demux in player, cosmetics) :)

FernetMenta replied:
> but why don't you just return m_pOtherStream != NULL;? 

not safe, open may fail which can result in this condition too

FernetMenta replied:
writing this I noticed that I set m_isOtherStreamHack at the wrong place. It needs to be set when trying to open otherStream

opdenkamp replied:
which case am I missing?

* no other stream: m_pOtherStream = NULL, m_isOtherStreamHack = false
* input stream failed to create: m_pOtherStream = NULL, m_isOtherStreamHack = false
* input stream failed to open: m_pOtherStream is deleted and NULLed, m_isOtherStreamHack = false
* other stream opened: m_pOtherStream != NULL, m_isOtherStreamHack = true

opdenkamp replied:
right, crosspost. just noticed that the results were all the same in the current code ;-)

opdenkamp commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStream.h:
```diff
@@ -77,6 +77,7 @@ class CDVDInputStream
     virtual bool Record(bool bOnOff) = 0;
     virtual bool CanPause() = 0;
     virtual bool CanSeek() = 0;
+    virtual bool IsOtherStreamHack() = 0;
```
looks like this needs a default `{ return false; }`?

opdenkamp replied:
since you've only implemented it for 1 inputstream

FernetMenta replied:
I intend to remove this generalization. IChannel is only used by PVR

opdenkamp replied:
right yeah, misleading github, didn't see that it was only in IChannel

FernetMenta replied:
this is a method of IChannel :)

```
virtual PVR::CPVRChannelPtr GetSelectedChannel() { return PVR::CPVRChannelPtr(); };
```

opdenkamp replied:
yeah github showed `class CDVDInputStream` above the the change, I'm stupid ;-)

FernetMenta replied:
you are not, we all stumbled into this already :)

FernetMenta committed:
VideoPlayer - PVR: demote IChannel interface to InputStreamPVRManager

FernetMenta commented:
@opdenkamp I moved assignment of m_isOtherStreamHack up and added a second commit which demotes IChannel interface.
Are you ok with leaving cosmetics in the first commit? Otherwise I would need to write that again :(

opdenkamp commented:
Sure, you're doing most of the work in /cores/ now anyway, so it's pretty much your choice ;-) I would have cleaned it up before merging though cause this ain't pretty if you need to bisect, revert, etc.

FernetMenta commented:
jenkins build this please
